File: lib/src/merge/merge_option.dart
import 'package:flutter/material.dart';

import '../convert_value.dart';
import '../image_source.dart';
import '../output_format.dart';
import '../position.dart';

/// Merge multiple images option.
class ImageMergeOption with JsonAble {
  ImageMergeOption({
    required this.canvasSize,
    this.format = const OutputFormat.jpeg(90),
  }) : assert(canvasSize > Size.zero);

  /// The canvas size. (output image size)
  final Size canvasSize;

  /// Output format for result image.
  final OutputFormat format;

  /// All merge image options.
  final List<MergeImageConfig> _mergeImageConfig = <MergeImageConfig>[];

  /// Add a merge image option.
  void addImage(MergeImageConfig config) {
    _mergeImageConfig.add(config);
  }

  @override
  Map<String, Object?> toJson() {
    return <String, Object?>{
      'images': _mergeImageConfig.map((e) => e.toJson()).toList(),
      'fmt': format.toJson(),
      'w': canvasSize.width.toInt(),
      'h': canvasSize.height.toInt(),
    };
  }
}

/// Merge image option.
class MergeImageConfig with JsonAble {
  /// Merge image option.
  const MergeImageConfig({
    required this.image,
    required this.position,
  });

  /// The image source.
  final ImageSource image;

  /// The position of image.
  final ImagePosition position;

  @override
  Map<String, Object?> toJson() {
    return <String, Object?>{
      'src': image.toJson(),
      'position': position.toJson(),
    };
  }
}

File: android/src/main/kotlin/com/fluttercandies/image_editor/option/MergeOption.kt
package com.fluttercandies.image_editor.option

class MergeOption(map: Map<*, *>) {
    val formatOption = FormatOption(map["fmt"] as Map<*, *>)
    val images: List<MergeImage>
    val width: Int = map["w"] as Int
    val height: Int = map["h"] as Int

    init {
        val list = ArrayList<MergeImage>()
        val src = map["images"] as List<*>
        for (any in src) {
            if (any is Map<*, *>) {
                list.add(MergeImage(any))
            }
        }
        images = list
    }
}

class MergeImage(map: Map<*, *>) {
    val byteArray: ByteArray = (map["src"] as Map<*, *>)["memory"] as ByteArray
    val position = ImagePosition(map["position"] as Map<*, *>)
}

class ImagePosition(map: Map<*, *>) {
    val x = map["x"] as Int
    val y = map["y"] as Int
    val width = map["w"] as Int
    val height = map["h"] as Int
}
